src/compiler/checker.ts

```diff
@@ -4919,7 +4919,7 @@ namespace ts {
             if (strictNullChecks && declaration.initializer && !(getFalsyFlags(checkDeclarationInitializer(declaration)) & TypeFlags.Undefined)) {
                 type = getTypeWithFacts(type, TypeFacts.NEUndefined);
             }
-            return declaration.initializer && !getContextualTypeForVariableLikeDeclaration(walkUpBindingElementsAndPatterns(declaration)) ?
+            return declaration.initializer && !getEffectiveTypeAnnotationNode(walkUpBindingElementsAndPatterns(declaration)) ?

```

Comment:
I not sure think this is right to change. Outside of strict mode, isn't this is going to make an `undefined` initializer (with a contextual type) be `any` instead of the correct contextual type.

Reply:
It's the right change (it basically just reverts the change that you had made). In classic checking mode this makes a variable of type `string` as expected.